utils/writeable/src/fallible.rs

```diff
@@ -0,0 +1,660 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use core::{convert::Infallible, fmt};
+
+use crate::*;
+
+/// The error returned while writing `Option<T>` when it is the `None` variant.
+#[derive(Debug)]
+pub struct MissingWriteableError;
+
+impl Writeable for MissingWriteableError {
+    #[inline]
+    fn write_to<W: fmt::Write + ?Sized>(&self, sink: &mut W) -> fmt::Result {
+        // Substitute with the Unicode replacement character
+        sink.write_char('\u{FFFD}')
+    }
+
+    #[inline]
+    fn writeable_length_hint(&self) -> LengthHint {
+        LengthHint::exact(3)
+    }
+}
+
+impl_display_with_writeable!(MissingWriteableError);
+
+#[test]
+fn test_missing_writeable_error() {
+    assert_writeable_eq!(MissingWriteableError, "�");
+}
+
+#[derive(Debug)]
+enum NeverWriteable {}
+
+impl Writeable for NeverWriteable {
+    #[inline(always)] // to help the compiler find unreachable code paths
+    fn write_to<W: fmt::Write + ?Sized>(&self, _sink: &mut W) -> fmt::Result {
+        match *self {}
+    }
+    #[inline(always)] // to help the compiler find unreachable code paths
+    fn write_to_parts<S: PartsWrite + ?Sized>(&self, _sink: &mut S) -> fmt::Result {
+        match *self {}
+    }
+    #[inline(always)] // to help the compiler find unreachable code paths
+    fn writeable_length_hint(&self) -> LengthHint {
+        match *self {}
+    }
+    #[inline(always)] // to help the compiler find unreachable code paths
+    fn write_to_string(&self) -> Cow<str> {
+        match *self {}
+    }
+    #[inline(always)] // to help the compiler find unreachable code paths
+    fn write_cmp_bytes(&self, _other: &[u8]) -> core::cmp::Ordering {
+        match *self {}
+    }
+}
+
+trait UnwrapInfallible {
+    type T;
+    fn unwrap_infallible(self) -> Self::T;
+}
+
+impl<T> UnwrapInfallible for Result<T, Infallible> {
+    type T = T;
+    #[inline(always)] // to help the compiler find unreachable code paths
+    fn unwrap_infallible(self) -> T {
+        match self {
+            Ok(t) => t,
+            Err(e) => match e {},
+        }
+    }
+}
+
+/// Lower-level trait for writeables that can fail while writing.
+///
+/// The methods on this trait should be implemented on custom writeables, but they generally
+/// should not called directly. For methods more useful in client code, see [`TryWriteable`].
+///
+/// # Implementor Notes
+///
+/// All functions take a `handler` argument. This argument is used by [`TryWriteable`] to
+/// customize the error handling behavior.
+///
+/// When your writeable implementation encounters an error, call the handler with an instance
+/// of your error type. The handler will either return `Ok` with a [`Writeable`] that should be
+/// written to the sink, or `Err` with an error that should be immediately returned up the stack.
```

Comment:
Why not pass a reference to the sink into the handler? It might want to do multiple writes, and then you'd have to introduce a new type in order to return a writeable.